### ISSUE TITLE
Rule proposal: prefer-tail-recursive-types

### ISSUE THREAD

**Issue by ronami** [typescript-eslint/typescript-eslint#10457]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

TypeScript v4.5 added support for [tail call optimization for conditional recursive types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-5.html#tail-recursion-elimination-on-conditional-types). Writing a recursive type in a tail-call optimized fashion makes it more performant, and also allows it to handle larger inputs ([playground link](https://typescript-eslint.io/play#ts=5.7.2&fileType=.tsx&code=C4TwDgpgBA4hwGEAWBDATgZwDwGUB8UAvFDlBAB7AQB2AJhlAAYAkA3gJbUBmEaUy6AL5tOPPgCUIGYIMYAoKFAD8-VHwA%2BseAMxZJ0vAqgAuKNQgA3XgG45c0JCgAVFOwA2cRGuw4ANFABBAGMgojNLXgJiUgoqOgYWDm5eVSERZIkpGXlFFRd3Tx1sfWB-HShNYKDDRVMq23twaABJMIAiJAg3NwB7KAB3HrQ3WiguHr6AI3QoaYAvOahOkCW%2Bt3gAcgYAcz6ARigAJigAZigAFigAVigANigAdigADigATigABja7AHpf5xNKCcaQoajAdgoCE9ajAhgUIJSDDsKxuFa0CAQMBQMGjMA9DDIyZo4HcTjsKgAOkajnEB2IhW8WGaeAa-ygAAlcesGMBOqSwABXYBQDCCkJIriC7ogGnQcTHYj5DzaJks2xAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)).

Recursive types may not be common, but they are invaluable in some cases. A rule that ensures they are written efficiently can be helpful, in my opinion.

### Fail Cases

```TypeScript
type GetChars<S> = S extends `${infer Char}${infer Rest}`
  ? Char | GetChars<Rest>
  : never;
```

### Pass Cases

```TypeScript
type TailGetChars<S, Acc = never> = S extends `${infer Char}${infer Rest}`
  ? TailGetChars<Rest, Char | Acc>
  : Acc;
```

### Additional Info

I think a rule like that may be especially useful for type challenges like [Type-Challenges](https://github.com/type-challenges/type-challenges) and [Type-Hero](https://github.com/typehero/typehero).

**Comment by JoshuaKGoldberg** [comment]
Love the idea, but how difficult is it to detect a type that could be switched to a tail-recursive form? My hunch is it would be quite difficult to get it right. Types can have a _lot_ of edge cases.

Nitpicking the name: I think we'd go with _`prefer-`_`tail-recursive-types` since it's not always doable.

**Comment by ronami** [comment]
That's a good question!

I've looked around to see if anything similar exists in other languages. The closet I could find is Scala's [`tailrec` annotation](https://www.scala-lang.org/api/current/scala/annotation/tailrec.html), which developers need to add for the compiler to verify that the function is in-fact tail-call.

A special comment can be supported (similar to Scala's `tailrec`) that developers can add so the rule will enforce the implementation is tail-call and remains tail-call throughout changes and refactors:

```ts
// @tseslint-tail-call
type GetChars<S> = S extends `${infer Char}${infer Rest}`
  ? Char | GetChars<Rest>
  : never;
```

This means the rule needs to be actively enabled by developers, which takes away some of its value.

---

Since TypeScript's type system is a lot more limited as a language than other languages (e.g. no stateful loops or trampolines), I think it may be feasible to have a set of heuristics to detect whether a recursive function can be refactored into a tail-call one.

In my opinion, even if the rule doesn't warn on every applicable recursive type but only on some, as long as it only reports ones that can be optimized, this is OK. Over time, the rule may be able to catch additional patterns, but it should never flag a generic that cannot be refactored into a tail call implementation.

Initially, the rule can flag recursive returns that have a single recursive call (under the same branch) and in which the result is used to create a new value:

```ts
type GetChars<S> = S extends `${infer Char}${infer Rest}`
  ? Char | GetChars<Rest>
  : never;

type GetCharsArray<S> = S extends `${infer Char}${infer Rest}`
  ? [Char, ...GetCharsArray<Rest>]
  : [];

type Filter1<T extends unknown[], P> = T extends [infer F, ...infer R]
  ? F extends P
    ? [F, ...Filter<R, P>]
    : Filter<R, P>
  : [];

type Filter2<T extends unknown[], P> = T extends [infer F, ...infer R]
  ? F extends P
    ? Concat<F, ...Filter<R, P>>
    : Filter<R, P>
  : [];
```

To my understanding, most recursive functions that cannot be optimized have multiple recursive calls under the same branch, and their result is combined (e.g. merge-sort or towers-of-hanoi):

```ts
type Hanoi<
  Rings extends number,
  FromRod extends string = 'A',
  ToRod extends string = 'B',
  IntermediateRod extends string = 'C',
  Acc extends 1[] = [],
> = Acc['length'] extends Rings
  ? []
  : [
      ...Hanoi<Rings, FromRod, IntermediateRod, ToRod, [...Acc, 1]>,
      [FromRod, ToRod],
      ...Hanoi<Rings, IntermediateRod, ToRod, FromRod, [...Acc, 1]>,
 ];
```

In addition, having a recursive call in a non-return branch should disqualify it as a tail-call candidate (like a conditional type's extends or check types). There are potentially (highly likely) edge cases I'm missing where this won't work and may require more looking into or adjustments.

---

Bottom line: I'm +1 for closing this as I'm not sure how practical creating such a rule is, at least until a more practical approach for tackling this is available.

**Comment by JoshuaKGoldberg** [comment]
👍 makes sense. Agreed we don't have the bandwidth to practically do this in typescript-eslint, as awesome as it would be.

If someone has the energy+time to write a community plugin we'd love to see that!
